### PR TITLE
Issue #5022 Filter Cache cleanup

### PR DESCRIPTION
**Pull request by gregw**
Issue #5022 Filter Cache cleanup:
 + Fixed many compiler warnings
 + removed old LazyList leftovers
 + Don't create holder string for source unless required
 + Only have a single type of chain, so it can be wrapped regardless of cache
 + Reverse mappings lists to make filter chain creation easier
 + build chain directly rather than build a list then a chain

Signed-off-by: Greg Wilkins <gregw@webtide.com>

### CODE REVIEW

File: jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java [hunk 1]
```diff
@@ -591,69 +592,69 @@ public void doHandle(String target, Request baseRequest, HttpServletRequest requ
...
-            for (FilterMapping filterPathMapping : _filterPathMappings)
-            {
-                if (filterPathMapping.appliesTo(pathInContext, dispatch))
-                    filters.add(filterPathMapping.getFilterHolder());
-            }
-        }
+        FilterChain chain = null;
 
-        // Servlet name filters
         if (servletHolder != null && _filterNameMappings != null && !_filterNameMappings.isEmpty())
```
Comment (janbartel) [inline jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java:611]: The Servlet Spec requires that the first filters in the chain are the url style mappings, and then the servlet name mappings. This does it the other way around, doesn't it?

**Review by janbartel — CHANGES_REQUESTED** [review]

File: jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java (hunk 1 above)
Comment (joakime) [reply jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java:611]: I would hope we fail some of our unit tests then.
We should have basic tests that validate these kinds of requirements from the servlet spec.
Unless Greg has done some kind of trickery and his "Reverse mappings" statement is really "mappings done in reverse" making the change he made valid again.

Comment (gregw) [reply jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java:611]: @janbartel The lists are in reverse so we apply the inner onion skin first. The order the filters are executed is not changed.

**Comment by joakime** [comment]
Test failure on "JDK8 / testHTTP2 – org.eclipse.jetty.osgi.test.TestJettyOSGiBootHTTP2" ?

**Review by janbartel — APPROVED** [review]

**Comment by gregw** [comment]
thanks @janbartel .... but as I was about to merge, I've got cold feet!!!!  I think this should be faster and better than the cache we currently have.... but I don't really know.  So I'm not going to merge for this release cycle.   @lorban @sbordet when you have time, we should chat about the possible performance impact of this change... it is on the main path to a Servlet and many big users have many many filters, so we need to get this right.

**Comment by gregw** [comment]
@lorban @sbordet can you find time to look at this and comment on possible performance impacts before I merge.

Previously we either build a one time filter chain class that used an iterator OR a cacheable chain that was put into both a the cache and a LRU list, so full cache could pop off single entries.

Now I build a filter chain the same way no matter what: a nested set of filter chain wrappers, that can be cached or used and discarded.  This avoids creating a temporary array, but there can be more wrappers.  The cache doesn't maintain an LRU and we just flush the entire cache if it fills.

**Comment by lorban** [comment]
I don't expect any perf impact from the lower-level changes, but these two higher-level changes may have some:
 - you doubled the default cache size
 - when the cache is full, it's dumped instead of trying to evict the oldest entry
but I don't know if that could have an impact in practice.

If those two don't bother you (I assume they don't) then this LGTM.

**Comment by gregw** [comment]
@lorban my thinking goes that for a webapp with few unique paths, then the cost of the LRU will be wasted as the cache will never fill up.
For apps with many unique paths it is a toss up between is it cheaper to remake the cache entirely vs keep a LRU cache... and that just depends so much on the URI usage pattern.

So I thought let's make it work as best we can for the many webapps that only have a few unique URIs and then double the cache size just to capture more webapps in that category.

**Pull request merged**